File: 2024/src/day01.fs
module Day01

open System

let parse =
    splitLines >> Array.map (split " " >> Array.map Int32.Parse) >> Array.transpose

let runPartOne input =
    let leftRight = parse input

    Array.map2 (fun a b -> a - b |> abs) (Array.sort leftRight[0]) (Array.sort leftRight[1])
    |> Array.sum

let runPartTwo input =
    let leftRight = parse input

    let counts =
        leftRight[1]
        |> Array.fold
            (fun map num ->
                map
                |> Map.change num (function
                    | Some value -> Some(value + num)
                    | None -> Some num))
            Map.empty

    leftRight[0]
    |> Array.sumBy (fun num -> Map.tryFind num counts |> Option.defaultValue 0)

File: 2024/src/day01.ts
function parse(input: string) {
  const left = [];
  const right = [];
  for (const line of input.split("\n")) {
    const values = line.split("   ").map(Number);
    left.push(values[0]);
    right.push(values[1]);
  }
  return [left, right];
}

export function runPartOne(input: string): number {
  const [left, right] = parse(input);
  left.sort();
  right.sort();
  return left.reduce((sum, num, i) => sum + Math.abs(num - right[i]), 0);
}

export function runPartTwo(input: string): number {
  const [left, right] = parse(input);
  const counts = right.reduce(
    (map, num) => map.set(num, (map.get(num) || 0) + num),
    new Map(),
  );
  return left.reduce((sum, num) => sum + (counts.get(num) || 0), 0);
}
